File: oxidize-golang/hf/hub.go
package hf

import (
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"os"
	"path/filepath"
	"sort"
	"strings"
	"time"
)

const defaultRevision = "main"

// ResolveOptions controls Hugging Face GGUF resolution.
type ResolveOptions struct {
	Repo       string
	Revision   string
	Filename   string
	CacheDir   string
	HTTPClient *http.Client
	APIBase    string
	CDNBase    string
}

// ResolveGGUF downloads (or returns a cached) GGUF file for a Hugging Face repo.
// Repo may be "org/model" or "org/model/quant.gguf" when Filename is empty.
func ResolveGGUF(opts ResolveOptions) (string, error) {
	repo, filename, err := splitRepoAndFile(opts.Repo, opts.Filename)
	if err != nil {
		return "", err
	}
	if opts.Filename != "" {
		filename = opts.Filename
	}
	if filename == "" {
		filename, err = pickSingleGGUF(repo, opts.Revision, opts.HTTPClient, opts.APIBase)
		if err != nil {
			return "", err
		}
	}
	rev := opts.Revision
	if rev == "" {
		rev = defaultRevision
	}
	cache := opts.CacheDir
	if cache == "" {
		cache = defaultCacheDir()
	}
	destDir := filepath.Join(cache, strings.ReplaceAll(repo, "/", "_"))
	if err := os.MkdirAll(destDir, 0o755); err != nil {
		return "", err
	}
	dest := filepath.Join(destDir, filename)
	if st, err := os.Stat(dest); err == nil && st.Size() > 0 {
		return dest, nil
	}
	tmp := dest + ".part"
	if err := downloadFile(resolveURL(opts.CDNBase, repo, rev, filename), tmp, opts.HTTPClient); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	if err := os.Rename(tmp, dest); err != nil {
		_ = os.Remove(tmp)
		return "", err
	}
	return dest, nil
}

func splitRepoAndFile(repo, explicitFile string) (string, string, error) {
	repo = strings.TrimSpace(repo)
	if repo == "" {
		return "", "", fmt.Errorf("hf: empty repo")
	}
	if explicitFile != "" {
		return repo, explicitFile, nil
	}
	parts := strings.Split(repo, "/")
	if len(parts) >= 3 && strings.HasSuffix(strings.ToLower(parts[len(parts)-1]), ".gguf") {
		file := parts[len(parts)-1]
		repo = strings.Join(parts[:len(parts)-1], "/")
		return repo, file, nil
	}
	return repo, "", nil
}

func defaultCacheDir() string {
	if home, err := os.UserHomeDir(); err == nil {
		return filepath.Join(home, ".cache", "oxidize", "hf")
	}
	return filepath.Join(os.TempDir(), "oxidize-hf")
}

func resolveURL(base, repo, revision, filename string) string {
	if base == "" {
		base = "https://huggingface.co"
	}
	return fmt.Sprintf("%s/%s/resolve/%s/%s", strings.TrimRight(base, "/"), repo, revision, filename)
}

type modelInfo struct {
	Siblings []struct {
		RFilename string `json:"rfilename"`
	} `json:"siblings"`
}

func pickSingleGGUF(repo, revision string, client *http.Client, apiBase string) (string, error) {
	names, err := ListGGUFFiles(repo, revision, client, apiBase)
	if err != nil {
		return "", err
	}
	switch len(names) {
	case 0:
		return "", fmt.Errorf("hf: repo %q has no .gguf files", repo)
	case 1:
		return names[0], nil
	default:
		var b strings.Builder
		b.WriteString(fmt.Sprintf("hf: repo %q has multiple .gguf files; specify --file. Candidates:\n", repo))
		for i, n := range names {
			if i >= 25 {
				b.WriteString("  ...\n")
				break
			}
			b.WriteString("  ")
			b.WriteString(n)
			b.WriteString("\n")
		}
		return "", fmt.Errorf("%s", strings.TrimSuffix(b.String(), "\n"))
	}
}

// ListGGUFFiles returns sorted .gguf filenames in a Hugging Face model repo.
func ListGGUFFiles(repo, revision string, client *http.Client, apiBase string) ([]string, error) {
	if revision == "" {
		revision = defaultRevision
	}
	if client == nil {
		client = &http.Client{Timeout: 2 * time.Minute}
	}
	if apiBase == "" {
		apiBase = "https://huggingface.co"
	}
	url := fmt.Sprintf("%s/api/models/%s", strings.TrimRight(apiBase, "/"), repo)
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	resp, err := client.Do(req)
	if err != nil {
		return nil, fmt.Errorf("hf: list %q: %w", repo, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 512))
		return nil, fmt.Errorf("hf: list %q: HTTP %d: %s", repo, resp.StatusCode, strings.TrimSpace(string(body)))
	}
	var info modelInfo
	if err := json.NewDecoder(resp.Body).Decode(&info); err != nil {
		return nil, fmt.Errorf("hf: decode model info: %w", err)
	}
	var out []string
	for _, s := range info.Siblings {
		if strings.HasSuffix(strings.ToLower(s.RFilename), ".gguf") {
			out = append(out, s.RFilename)
		}
	}
	sort.Strings(out)
	return out, nil
}

func downloadFile(url, dest string, client *http.Client) error {
	if client == nil {
		client = &http.Client{Timeout: 0}
	}
	req, err := http.NewRequest(http.MethodGet, url, nil)
	if err != nil {
		return err
	}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("hf: download %s: %w", url, err)
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return fmt.Errorf("hf: download %s: HTTP %d", url, resp.StatusCode)
	}
	f, err := os.Create(dest)
	if err != nil {
		return err
	}
	defer f.Close()
	if _, err := io.Copy(f, resp.Body); err != nil {
		return err
	}
	return f.Close()
}

File: oxidize-python/oxidize_python/hf/hub.py
"""Hugging Face GGUF resolver mirroring oxidize-golang/hf/hub.go."""

from __future__ import annotations

import json
import urllib.request
from dataclasses import dataclass
from pathlib import Path

DEFAULT_REVISION = "main"


@dataclass
class ResolveOptions:
    repo: str
    revision: str = DEFAULT_REVISION
    filename: str = ""
    cache_dir: str = ""
    api_base: str = "https://huggingface.co"
    cdn_base: str = "https://huggingface.co"


def _default_cache_dir() -> Path:
    home = Path.home()
    return home / ".cache" / "oxidize" / "hf"


def _split_repo_and_file(repo: str, explicit_file: str) -> tuple[str, str]:
    repo = repo.strip()
    if not repo:
        raise ValueError("hf: empty repo")
    if explicit_file:
        return repo, explicit_file
    parts = repo.split("/")
    if len(parts) >= 3 and parts[-1].lower().endswith(".gguf"):
        return "/".join(parts[:-1]), parts[-1]
    return repo, ""


def list_gguf_files(
    repo: str,
    revision: str = DEFAULT_REVISION,
    api_base: str = "https://huggingface.co",
) -> list[str]:
    _ = revision
    url = f"{api_base.rstrip('/')}/api/models/{repo}"
    req = urllib.request.Request(url, method="GET")
    with urllib.request.urlopen(req, timeout=120) as resp:
        payload = json.loads(resp.read().decode())
    names: list[str] = []
    for sibling in payload.get("siblings", []):
        name = sibling.get("rfilename", "")
        if name.lower().endswith(".gguf"):
            names.append(name)
    names.sort()
    return names


def _pick_single_gguf(repo: str, revision: str, api_base: str) -> str:
    names = list_gguf_files(repo, revision, api_base)
    if not names:
        raise ValueError(f"hf: repo {repo!r} has no .gguf files")
    if len(names) == 1:
        return names[0]
    lines = "\n".join(f"  {n}" for n in names[:25])
    if len(names) > 25:
        lines += "\n  ..."
    raise ValueError(
        f"hf: repo {repo!r} has multiple .gguf files; specify --file. Candidates:\n{lines}"
    )


def _download(url: str, dest: Path) -> None:
    dest.parent.mkdir(parents=True, exist_ok=True)
    tmp = dest.with_suffix(dest.suffix + ".part")
    req = urllib.request.Request(url, method="GET")
    with urllib.request.urlopen(req, timeout=0) as resp, tmp.open("wb") as out:
        while True:
            chunk = resp.read(1024 * 1024)
            if not chunk:
                break
            out.write(chunk)
    tmp.replace(dest)


def resolve_gguf(opts: ResolveOptions) -> str:
    repo, filename = _split_repo_and_file(opts.repo, opts.filename)
    if opts.filename:
        filename = opts.filename
    if not filename:
        filename = _pick_single_gguf(repo, opts.revision or DEFAULT_REVISION, opts.api_base)
    rev = opts.revision or DEFAULT_REVISION
    cache = Path(opts.cache_dir) if opts.cache_dir else _default_cache_dir()
    dest_dir = cache / repo.replace("/", "_")
    dest_dir.mkdir(parents=True, exist_ok=True)
    dest = dest_dir / filename
    if dest.is_file() and dest.stat().st_size > 0:
        return str(dest)
    url = f"{opts.cdn_base.rstrip('/')}/{repo}/resolve/{rev}/{filename}"
    _download(url, dest)
    return str(dest)
